### PR TITLE
Add particle filters to picmi

### PR DESCRIPTION
**Pull request by chillenzer**
- [x] Rebase after #5606 is merged. Changes on top of #5606 start from `0f9664aec226807bc5836856f81847e5b5b14a58` "Add register_codename call"

## Add Particle Filters to PICMI
This PR allows to use `FilteredSpecies` instead of `Species` wherever PIConGPU allows it. Its implementation is rather messy for reasons laid out below but it is well-tested and seems to work fine.

### User Interface
The general user interface is (with the most nonsensical filter I could come up with): 
```python
electrons = Species(name='electrons')
heavy_electrons = FilteredSpecies(
            species=electrons,
            functor=ParticleFilter(name="massFilter", functor=lambda particle: particle.get("mass") > 0.5),
        )
```
This can be used in all diagnostics (for which PIConGPU supports filters) as well as Collisions. Radiation came late to the party and is not yet implemented.

Particle filters can use random numbers as follows:
```python
def sample_half(particle, rng: RNGArg):
    # defaults to range=(0,1), return_type='float_X'
    return rng.get('uniform') < 0.5

FilteredSpecies(species=electrons, functor=ParticleFilter(name='sampling', functor=sample_half))
```

The `RNGArg` class is self-aware in the sense that it can translate itself into a `scipy.stats` distribution which can be quite convenient in some situations:
```python
from functools import partial

# The type annotation of `RNGArg` is mandatory!
def sample(particle, rng: RNGArg, percent):
    """Sample `percent`% of the particles"""
    # Why on earth anybody would use a normal distribution for this is a different questions:
    distribution = {'dist': 'normal', 'std': 5.0, 'mean': 42.0}
    return rng.get(**distribution) < rng.to_scipy(**distribution).ppf(percent / 100)
    
# Randomly draws 10% of the species each time:
FilteredSpecies(..., functor=ParticleFilter(..., functor=partial(sample, percent=10)))
```
Due to limitations on the PIConGPU side, we can currently only draw from a single type of distribution (`uniform` or `normal`) within one functor but there are plans to lift this. `range`, `std` and `mean` can be changed within one functor.

### Tests
This feature is well tested in the `end-to-end` tests. Please run them! I've implemented position filters comparing against `range`-restricted (but unfiltered) particle dumps and I've checked that the number of randomly `sample`d particles (with a filter similar to the above) matches the percent of particles expected.

### Implementation details
There are three major complications that make this a mess on the implementation side of things:
- The PIConGPU native C++ interface has (at least) 4 different ways to accept particle filters in different situations. This means that each template is more-or-less hand-rolled for that specific `.param` file. That, in turn, implies that as my implementation approached feature parity I added new information to the rendering context as necessary. I don't claim that the templates are written in a consistent style or that the PyPIConGPU interface towards the backend caters information in a consistent way. But I believe that it makes more sense to refactor this once we have the whole picture (and the whole list of requirements) and potentially also touch the C++ side of things to allow more uniform code generation.
- All kinds of particle functors have different capabilities. So, I can't just generate one set of statements per particle functor and use that everywhere but instead I have to respect the context in which a functor is used and translate into different C++ statements depending on how that functor is supposed to be used later.
- There's a general design problem in our PICMI-PyPIConGPU translation: The `.get_as_pypicongpu` interface bears the implicit assumption that there is a one-to-one mapping of PICMI entities to PyPIConGPU entities. A good example in this context is that `FilteredSpecies` just don't exist as a concept in PIConGPU (except for the binning plugin). Instead `FilteredSpecies` split up into a particle filter, a species and a relation between the two that appears in a particular context and has different incarnations depending on the context. Thus, `FilteredSpecies.get_as_pypicongpu()` cannot follow the envisioned protocol of returning a single-object PyPIConGPU representation. The problem is quite obvious: PICMI is a declarative language and the implementation of a declarative language is a constraint solver. That is more-or-less precisely what I've recently implemented for `Species` specifically already but we need it for the full translation of a `Simulation`. Any piece of information on the PICMI level imposes constraints on potentially any piece of information on the PyPIConGPU level and it's the task of a constraint solver to construct a valid `pypicongpu.Simulation` that abides by all constraints. You might realise at this point that another major refactoring is dawning upon us! But for now, I've written some messy ad-hoc solution gather particle filters from where I know they can appear. Not great but it works for now.

Furthermore, I've taken a new approach to selecting different blocks in a template: `type_filtered` (always set to `True`) is an element of the `FilteredSpecies` context but not the bare `Species` context. Doing `{{#type_filtered}}` results in the block being rendered in case of a `FilteredSpecies` but not for bare `Species` which is precisely what we want. Unfortunately, with the current set up a missing key results in a lot of warning. My plan is to switch to another mustache-rendering library that will allow us to selectively disable this warning for specific tags that we use like this. This new system will allow us to get rid of all the `SelfRegistering` shenanigans.

### Other changes
I've sneaked in a few minor refactorings/changes:
- I added a call to `PICMI_standard.register_codename` to our main `__init__.py`. I don't think it changes anything but it's what the PICMI standard protocol suggests to do.
- I've added a `ValueError` to `_apply_units` for `RangeSpec`. Unrelated but useful.
- I've added the `alt` function, basically an inline `try-except` statement, and it is spreading quickly throughout the codebase.
- I've removed the custom user input section from the example. It is unusable because we don't serve the corresponding custom templates and it diminished significantly in functionality since we took a big leap towards feature parity.

### Status and review
I've squashed history enough that individual features constitute single commits. One could squash further, please speak up if you'd prefer me to do so. That's a quite non-invasive process. I wouldn't want to re-arrange history in a way that could introduce bugs though.

For review, please be aware that some files have "simply" been moved. In particular, functionality related to particle functors moved to a separate submodule. Make sure to use a proper diff command that actually shows you the differences like
```bash
git diff $(git merge-base mainline/dev ???) --color-moved --color-moved-ws=ignore-all-space
```
The various "Add filtered species to..." commits are probably good to review individually because you can clearly follow the flow of information through the interfaces and you can mostly focus on the templates that implement such functionality in the backend.

### CODE REVIEW

File: lib/python/picongpu/picmi/simulation.py [hunk 1]
```diff
@@ -392,6 +392,40 @@ def _check_compatibility(self):
         if self.max_steps is None and self.max_time is None:
             raise ValueError("runtime not specified (neither as step count nor max time)")
 
+    def _collect_particle_filters(self):
+        def extract_filtered_species(diagnostics):
+            return map(
+                lambda d: d.species,
+                filter(lambda d: alt(lambda: isinstance(d.species, FilteredSpecies), False), diagnostics),
+            )
+
```
Comment (ikbuibui) [inline lib/python/picongpu/picmi/simulation.py]: Looks unused
```suggestion
```

Comment (chillenzer) [reply lib/python/picongpu/picmi/simulation.py]: Addressed in the most non-trivial way I could think of. ;D

File: test/python/picongpu/end-to-end/diagnostics.py [hunk 2]
```diff
@@ -138,6 +191,62 @@ def generate_derived_field_dumps(species, functors):
...
+def _name_of(distribution):
+    # CAUTION: stackoverflow suggests that this might be the better way
+    # (mostly because of the efficiency of hashing vs. sorting):
+    #
+    #   unique_name = compute_hash(str(frozenset(distribution.items())).encode()).hexdigest()
+    #
+    # But it turns out that the above does not yield reproducible results between different runs.
+    # Probing the following a few times on my local machine
+    # found it to be sufficiently reproducible to work with conveniently:
+    unique_name = compute_hash(str(tuple(sorted(distribution.items()))).encode()).hexdigest()
```
Comment (ikbuibui) [inline test/python/picongpu/end-to-end/diagnostics.py:231]: If the floats representing the same number are different in different instances of the same distribution (idk how likely it is that this can happen), then you will have unique hashes.

**Review by ikbuibui — APPROVED** [review]
Looks fine. I rely on the tests to give me more confidence that things are okay.
I am not sure how to feel about `alt`. It seems both brilliantly hacky and something which should not exist. 
I feel similarly uneasy about `UnpackChain`.
I don't love that `Species` had to be replaced with `Species | Filtered Speices` everywhere. Maybe there is a cleaner way to extend things and require an interface. Or maybe this is an absolutely un-pythonic way of thinking. 
I think I am not the right person to review this stuff. In general I am not able to come to any good judgements on the maintainability of the patterns used.

I will merge this tomorrow around noon in case you want to comment on something (not really required).

File: test/python/picongpu/end-to-end/diagnostics.py (hunk 2 above)
Comment (chillenzer) [reply test/python/picongpu/end-to-end/diagnostics.py:231]: It's fine, I guess. It's supposed to be a "good enough" solution. The only purpose is to be able to re-use the simulation results while working on the tests (and obviously to yield unique names for C++). If the odd edge case has you re-compiling a test occasionally, that doesn't hurt.

**Comment by chillenzer** [comment]
> Looks fine. I rely on the tests to give me more confidence that things are okay.

Yes, I'm quite confident that the parts that are covered are covered very well. Haven't gone for exhaustiveness, though, but we've got people who'll certainly find all the edge cases (and a few more...)

> I am not sure how to feel about `alt`. It seems both brilliantly hacky and something which should not exist. I feel similarly uneasy about `UnpackChain`.

Yes, they are the kind of useful tool that Voodoo is. I'm sure you wouldn't want to let go of it once you've got it, but it is certainly questionable to have it in the first place. I think the difficulty is that it's not a native language idiom: It reads very well if you are familiar with its semantics but not being natively available means that few people are.

> I don't love that `Species` had to be replaced with `Species | Filtered Speices` everywhere. Maybe there is a cleaner way to extend things and require an interface. Or maybe this is an absolutely un-pythonic way of thinking.

Yes, if that's a problem, one could easily define `AnySpecies = Species | FilteredSpecies` and have `species: AnySpecies` everywhere. That's what's done for lasers, distributions, etc. already. I've refrained from doing it yet because
- a set of two is not too inconvenient to write out everywhere
- there are a few cases where only one (typically `Species`) is allowed
- the upcoming introduction of `picmi.MultiSpecies` would introduce more of such edge cases.

> I think I am not the right person to review this stuff. In general I am not able to come to any good judgements on the maintainability of the patterns used.

That is indeed an unfortunate thing. We should bring this up in the dev meeting again.

> I will merge this tomorrow around noon in case you want to comment on something (not really required).

:+1:
